sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs

```diff
@@ -19,161 +19,183 @@
             ImmutableArray<TLanguageKindEnum> syntaxKinds)
-            where TLanguageKindEnum : struct
-        {
+            where TLanguageKindEnum : struct =>
             RegisterSyntaxNodeAction(action, syntaxKinds.ToArray());
-        }
 
         internal void RegisterSyntaxNodeAction<TLanguageKindEnum>(Action<SyntaxNodeAnalysisContext> action,
             params TLanguageKindEnum[] syntaxKinds)
-            where TLanguageKindEnum : struct
-        {
+            where TLanguageKindEnum : struct =>
             context.RegisterSyntaxNodeAction(
                 c =>
                 {
-                    if (IsAnalysisDisabled(c.Node.SyntaxTree))
+                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
                     {
-                        return;
+                        action(c);
                     }
+                }, syntaxKinds);
 
-                    action(c);
-                },
-                syntaxKinds);
-        }
-
-        internal void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext> action)
-        {
+        internal void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext> action) =>
             context.RegisterSyntaxTreeAction(
                 c =>
                 {
-                    if (IsAnalysisDisabled(c.Tree))
+                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
                     {
-                        return;
+                        action(c);
                     }
-
-                    action(c);
                 });
-        }
 
-        internal void RegisterSymbolAction(Action<SymbolAnalysisContext> action, ImmutableArray<SymbolKind> symbolKinds)
-        {
+        internal void RegisterSymbolAction(Action<SymbolAnalysisContext> action, ImmutableArray<SymbolKind> symbolKinds) =>
             RegisterSymbolAction(action, symbolKinds.ToArray());
-        }
 
-        public void RegisterSymbolAction(Action<SymbolAnalysisContext> action, params SymbolKind[] symbolKinds)
-        {
+        public void RegisterSymbolAction(Action<SymbolAnalysisContext> action, params SymbolKind[] symbolKinds) =>
             context.RegisterSymbolAction(
                 c =>
                 {
-                    if (IsAnalysisDisabled(c.Symbol.Locations.FirstOrDefault(l => l.SourceTree != null)?.SourceTree))
+                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
                     {
-                        return;
+                        action(c);
                     }
+                }, symbolKinds);
 
-                    action(c);
-                },
-                symbolKinds);
-        }
-
-        internal static bool IsAnalysisDisabled(SyntaxTree tree)
+        internal static bool IsAnalysisDisabled(SyntaxTree tree, IEnumerable<DiagnosticDescriptor> supportedDiagnostics = null)

```

Comment:
I generally prefer positive conditions and IsAnalysisEnabled sounds better to me (and is easier to understand when there is negation). Not sure if this should be changed though.

Reply:
This one could be changed but weirdly enough when negating all conditions some tests are failing and I couldn't really understand what was going on.